src/pipdeptree/_discovery.py

```diff
@@ -2,28 +2,71 @@
 
 import site
 import sys
-from importlib.metadata import distributions
-from typing import TYPE_CHECKING
+from importlib.metadata import Distribution, distributions
 
-if TYPE_CHECKING:
-    from importlib.metadata import Distribution
+from pipdeptree._util import pep503_normalize
 
 
 def get_installed_distributions(
     local_only: bool = False,  # noqa: FBT001, FBT002
     user_only: bool = False,  # noqa: FBT001, FBT002
 ) -> list[Distribution]:
-    if user_only:
-        return list(distributions(path=[site.getusersitepackages()]))
-
-    # NOTE: See https://docs.python.org/3/library/venv.html#how-venvs-work for more details.
+    # See https://docs.python.org/3/library/venv.html#how-venvs-work for more details.
     in_venv = sys.prefix != sys.base_prefix
+    orginal_dists = []
 
     if local_only and in_venv:
         venv_site_packages = site.getsitepackages([sys.prefix])
-        return list(distributions(path=venv_site_packages))
+        orginal_dists = distributions(path=venv_site_packages)
+    elif user_only:
+        orginal_dists = distributions(path=[site.getusersitepackages()])
+    else:
+        orginal_dists = distributions()
+
+    # Since importlib.metadata.distributions() can return duplicate packages, we need to handle this. pip's approach is
+    # to keep track of each package metadata it finds, and if it encounters one again it will simply just ignore it. We
+    # take it one step further and warn the user that there are duplicate packages in their environment.
+    # See https://github.com/pypa/pip/blob/7c49d06ea4be4635561f16a524e3842817d1169a/src/pip/_internal/metadata/importlib/_envs.py#L34
+    seen_dists: dict[str, Distribution] = {}
+    first_seen_to_already_seen_dists_dict: dict[Distribution, list[Distribution]] = {}
+    dists = []
+    for dist in orginal_dists:
+        normalized_name = pep503_normalize(dist.metadata["Name"])
+        if normalized_name not in seen_dists:
+            seen_dists[normalized_name] = dist
+            dists.append(dist)
+            continue
+        already_seen_dists = first_seen_to_already_seen_dists_dict.setdefault(seen_dists[normalized_name], [])
+        already_seen_dists.append(dist)
+
+    if first_seen_to_already_seen_dists_dict:
+        render_duplicated_dist_metadata_text(first_seen_to_already_seen_dists_dict)
+
+    return dists
+
+
+FirstSeenWithDistsPair = tuple[Distribution, Distribution]
+
+
+def render_duplicated_dist_metadata_text(
```

Comment:
It is better to place this function to in `_render` directory.

Reply:
I think it makes sense to have the printing (or as we call rendering) logic in its own package, but we have also been doing this in `_validate.py` for some time so I think we can ponder this in a refactoring PR in the future